Jamesrowan/course subject integration

jrowan opened the pull request:
Adds pipeline functionality to pull the course subjects information from drupal, store that information on s3, and then load that information into the Vertica data warehouse.  The Vertica luigi target class and Vertica bulkloading luigi task class, modeled off the corresponding mysql tasks, should be usable for other data warehouse loading tasks as well.

@jab5569 @brianhw @mulby

brianhw commented on client.cfg:
```diff
@@ -1,34 +1,32 @@
+source = s3://<bucket to where all tracking logs are synched>/tracking/
+
+[hive]
+warehouse_path = http://s3.amazonaws.com/edx-analytics-scratch/jrowan/hive
```
I would respecfully suggest that this be obfuscated like the other S3 paths. :)

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+"""Collect the course catalog from Drupal for processing of course metadata like subjects or types"""
```
I would strongly suggest that we not use "Drupal" in this file, but rather use the "Course Catalog API".

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+from edx.analytics.tasks.util.overwrite import OverwriteOutputMixin
+from edx.analytics.tasks.util.vertica_load import VerticaCopyTask, VerticaCopyTaskMixin
+
+import urllib3.contrib.pyopenssl
```
In the cybersource and paypal code, we added the following comment:
# Tell urllib3 to switch the ssl backend to PyOpenSSL.
# see https://urllib3.readthedocs.org/en/latest/security.html#pyopenssl

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+    Pulls are made daily to keep a full historical record.
+    """
+
+    CATALOG_URL = 'https://www.edx.org/api/catalog/v2/courses'
```
This should be a parameter that is pulled from the configuration file -- at least the root of it.  OpenEdx installations should not be pulling from edx.org.

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+
+    CATALOG_URL = 'https://www.edx.org/api/catalog/v2/courses'
+    CATALOG_FORMAT = 'json'
+    run_date = luigi.DateParameter(default=datetime.date.today())
```
A small but sometimes significant detail:  today() returns the date in local time.  This should probably be datetime.datetime.utcnow().date() for general UTC consistency.

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+class PullFromDrupalMixin(OverwriteOutputMixin, WarehouseMixin):
+    """Define common parameters for Drupal course catalog pull and downstream tasks."""
+
+    run_date = luigi.DateParameter(default=datetime.date.today())
```
A small but sometimes significant detail:  today() returns the date in local time.  This should probably be datetime.datetime.utcnow().date() for general UTC consistency.

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+        """Output is in the form {warehouse_path}/catalog/{CCYY-mm-dd}/catalog.json"""
+        date_string = self.run_date.strftime('%Y%m%d')
+        filename = "catalog.{}".format(self.CATALOG_FORMAT)
+        url_with_filename = url_path_join(self.warehouse_path, "catalog", date_string, filename)
```
I would probably suggest a longer name than just "catalog" here.  Things are getting pretty clogged at root level of warehouse_path.  We might want to add more hierarchy at some point, but I'd lengthen this to "course_catalog_api" or something similar.

mulby commented on client.cfg:
```diff
@@ -1,34 +1,32 @@
-[hadoop]
```
You should revert all changes to this file. In your local environment you can just create an "override.cfg" file in the root of the repo that will be used by luigi.

This file actually contains the defaults for some variables used by luigi.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+
+    def output(self):
+        """Output is in the form {warehouse_path}/catalog/{CCYY-mm-dd}/catalog.json"""
+        date_string = self.run_date.strftime('%Y%m%d')
```
We typically use the ISO8601 date format:

%Y-%m-%d

Or more conveniently: `self.run_date.isoformat()`

mulby replied:
Also we follow the hive partition convention for separating snapshots of things like this by prefixing the date with "dt=". If/when we ever use this as a hive table it will prefer the folder hierarchy to be in this format.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+    def output(self):
+        """Output is in the form {warehouse_path}/catalog/{CCYY-mm-dd}/catalog.json"""
+        date_string = self.run_date.strftime('%Y%m%d')
+        filename = "catalog.{}".format(self.CATALOG_FORMAT)
```
I'm OK with harcoding ".json" here, I think the constant is unnecessary unless it's actually being passed to the API or something in a "Accept" header or something similar.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+
+class DailyProcessFromCatalogSubjectTask(PullFromDrupalMixin, luigi.Task):
+    """
+    A task that reads a local file generated from a daily catalog pull, and writes the course id and subject to a csv.
```
Update this to indicate that you are generating a tsv instead of a csv

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+                    if subjects is None or len(subjects) == 0:
+                        line = [
+                            course_id.encode('utf-8'),
+                            self.run_date,
```
I would be explicit here about the string format you want for the date.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+                        line = [
+                            course_id.encode('utf-8'),
+                            self.run_date,
+                            '\N'.encode('utf-8'),  # pylint: disable-msg=W1402
```
you may want to move the encoding down into a list comprehension, something like:

```python
output_file.write('\t'.join([v.encode('utf-8') for v in line]))
```

It's a little more DRY.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+                    # This will be a list of dictionaries with keys 'title', 'uri', and 'language'.
+                    # Encode in utf-8 as in general the subjects could be given in other languages.
+                    subjects = course.get('subjects', None)
+                    # It's possible no subjects are given for the course, in which case we record the lack of subjects.
```
Do we need to record this lack of subjects?

When joining with this table we can just use a left outer join and get null values.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+
+    @property
+    def table(self):
+        return "experimental.d_course_subjects"
```
Can we push the schema name into the config file and/or a parameter? Similarly to how we define the hive database we are writing to and the LMS database we are reading from?

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,173 @@
+        return "experimental.d_course_subjects"
+
+    @property
+    def auto_primary_key(self):
```
Consider not making these properties required if they are commonly going to just be `None` (same for default columns).

mulby replied:
Ah, I see you are overriding the default behavior... why do you want to do that here? Maybe explain in a comment?

mulby commented on edx/analytics/tasks/util/vertica_load.py:
```diff
@@ -0,0 +1,275 @@
+"""
```
I thought we were going to commit this to the edx/luigi fork?

https://github.com/edx/luigi

mulby replied:
Oh, wrong file, I meant to put this comment near the vertica target.

mulby commented on edx/analytics/tasks/util/vertica_load.py:
```diff
@@ -0,0 +1,275 @@
+        using the same transaction.
+        """
+        # In Vertica, tables are specified as schema_name.table_name, so just the first part of the table is needed.
+        schema = self.table.split('.')[0]
```
why not `self.database`?

jrowan replied:
@mulby I'm changing this to using self.schema and removing self.database.  I'm not using self.database to refer to the schema since the pyODBC/vertica-python end of the target thinks database refers to a specific other concept related to making connections.

mulby commented on edx/analytics/tasks/util/vertica_load.py:
```diff
@@ -0,0 +1,275 @@
+
+    def output(self):
+        """
+        Returns a MysqlTarget representing the inserted dataset.
```
s/Mysql/Vertica/

mulby commented on edx/analytics/tasks/util/vertica_load.py:
```diff
@@ -0,0 +1,275 @@
+            # first clear the appropriate rows from the luigi Vertica marker table
+            marker_table = self.output().marker_table  # side-effect: sets self.output_target if it's None
+            try:
+                query = "DELETE FROM {marker_table} where `target_table`='{target_table}'".format(
```
Not sure what the vertica row locking behavior is, but we have to be very careful to only lock the row we are deleting here. Mysql was locking the whole table which was preventing multiple jobs from being able to write into it at once.

mulby commented on edx/analytics/tasks/util/vertica_load.py:
```diff
@@ -0,0 +1,275 @@
+            query = "DELETE FROM {table}".format(table=self.table)
+            connection.cursor().execute(query)
+
+    def copy_data_table_from_target(self, cursor, delimiter="E'\t'", null_sequence="'\\N'"):
```
This is very cool. We should make the delimeter and null_sequence easier to override, though, maybe as properties that a subclass can override?

I can imagine wanting to read a variety of different file formats that have different conventions for these strings.

mulby commented:
Finished my pass. Great job.

Blockers for me:
* revert client.cfg
* make use of the `database` parameter to specify the schema instead of using "schema.table_name"

jrowan commented:
Unit and acceptance tests added, client.cfg reverted, and the schema to use is a configurable parameter now.  

@mulby @brianhw @jab5569

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+        date_string = "dt=" + self.run_date.strftime('%Y-%m-%d')  # pylint: disable=no-member
+        url_with_filename = url_path_join(self.warehouse_path, "course_catalog_api/catalog",
+                                          date_string, "catalog.json")
+        print "URL_WITH_FILENAME: ", str(url_with_filename)
```
Remove debugging print statements (throughout this PR)

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+    def output(self):
+        """Output is in the form {warehouse_path}/course_catalog_api/catalog/{CCYY-mm-dd}/catalog.json"""
+        date_string = "dt=" + self.run_date.strftime('%Y-%m-%d')  # pylint: disable=no-member
+        url_with_filename = url_path_join(self.warehouse_path, "course_catalog_api/catalog",
```
I would prefer this let `url_path_join` handle all of the separators here

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+    """Define common parameters for the course catalog API pull and downstream tasks."""
+
+    run_date = luigi.DateParameter(default=datetime.datetime.utcnow().date())
+    catalog_url = luigi.Parameter(default_from_config={'section': 'course-catalog', 'name': 'catalog_path'})
```
I would prefer the variable name and the config file parameter name were the same

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+            output_file.write(response.content)
+
+    def output(self):
+        """Output is in the form {warehouse_path}/course_catalog_api/catalog/{CCYY-mm-dd}/catalog.json"""
```
This comment looks like it's out of date

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+
+    def run(self):
+        date_string = "dt=" + self.run_date.strftime('%Y-%m-%d')  # pylint: disable=no-member
+        url_with_filename = url_path_join(self.warehouse_path, "course_catalog_api/catalog",
```
I think I would prefer simply "course_catalog" instead of "course_catalog_api" since the fact that it is pulled from an API is an implementation detail... in the future we might pull it directly from a database (for example).

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,202 @@
+        with self.input().open('r') as input_file:
+            # Since the course catalog is of fairly manageable size, we can read it all into memory at once.
+            # If this needs to change, we should be able to parse the catalog course by course.
+            catalog = json.loads(input_file.read()).get('items', None)
```
`None` is the default return value from `dict.get(key)`, so you do not need to explicitly specify it here.

jrowan commented:
I performed some modifications to the bulkloader based on infrastructure from the events-to-vertica task

jrowan commented:
Note that there are updates to the Vertica bulkloading infrastructure that I still need to push here, as we are now using a newer version of vertica-python (forcing function on that was the event loading).

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,198 @@
+        date_string = self.run_date.strftime('%Y-%m-%d')  # pylint: disable=no-member
+        partition_path_spec = HivePartition('dt', date_string).path_spec
+        filename = "subjects.tsv"
+        url_with_filename = url_path_join(self.warehouse_path, "course_catalog_api/subjects",
```
I think I put this in previous comments, but I two notes here:

1) I think we should call the folder simply "course_catalog". I don't think the "_api" suffix provides valuable information
2) I would prefer this was simply: `url_path_join(self.warehouse_path, 'course_catalog', 'subjects', ...)` since it keeps all separator logic in `url_path_join`.

mulby commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,198 @@
+        )
+
+
+class CourseCatalogWorkflowFixed(PullCatalogMixin, VerticaCopyTaskMixin, luigi.WrapperTask):
```
This looks duplicated?

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -69,7 +71,9 @@ def setUp(self):
 
+        self.catalog_path = 'http://acceptance.test/api/courses/v2'
         database_name = 'test_' + self.identifier
+        schema = 'test'
```
This should incorporate `self.identifier` to ensure that multiple tests can be run at the same time, I would recommend changing this to `'test_' + self.identifier`.

mulby replied:
Or you could simply change the constructor call for the `VerticaService` to take in the `database_name` value instead of `schema`.

mulby commented on edx/analytics/tasks/tests/acceptance/services/task.py:
```diff
@@ -72,6 +72,8 @@ def delete_existing_logs(self):
     def override_config(self, config_parser, overrides):
         for section_name, section in overrides.iteritems():
+            print section_name
+            print section
```
Remove debug print statements

jab5569 replied:
Do we need to remove these print statments?

mulby commented on edx/analytics/tasks/tests/acceptance/services/vertica.py:
```diff
@@ -0,0 +1,70 @@
+from edx.analytics.tasks.url import get_target_from_url
+
+
+class VerticaService(object):
```
This looks very similar to the DatabaseService. I would recommend creating a common base class that implements the shared logic.

We may be able to just override the `connect()` method, the `reset()` method and credentials url and have the rest be common.

jrowan replied:
I thought about this, but I think there's a sufficiently breaking difference between MySQL and Vertica to warrant not doing this.  My reasoning:

MySQL and Vertica mean very different things by "database," (MySQL databases are roughly equivalent to Vertica schemata, and the vertica connection module uses database as a kwarg but doesn't actually need it to make a connection), and there is a set of implicit_db logic in the DatabaseService that can't be used in Vertica.  If we were to make a base class, we'd have to override almost all of the logic in both places anyways.

mulby replied:
Got it, sounds reasonable.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+
+        assert 'oddjob_jar' in self.config
+
+        self.oddjob_jar = self.config['oddjob_jar']
```
is this actually used anywhere?

jrowan replied:
What are you referring to by "this"? There are 15 lines of code highlighted here.

mulby replied:
Sorry - could have been more clear here. Are you using `self.oddjob_jar`?

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+    def upload_data(self):
+        """Puts the test course catalog where the processing task would look for it, bypassing calling the actual ALI"""
+        src = os.path.join(self.data_dir, 'input', self.INPUT_FILE)
+        dst = url_path_join(self.warehouse_path, 'course_catalog_api/catalog/dt=2015-06-29', self.INPUT_FILE)
```
I would prefer to have `url_path_join` handle the separators here

jrowan replied:
Fixed with respect to hardcoding the slashes in 'course_catalog_api/catalog/dt=2015-06-29'; I have os.path.join for the src because those files are local to the machine running the acceptance test.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+
+        self.task.launch([
+            'CourseCatalogWorkflow',
+            '--credentials', self.vertica.vertica_creds_url,
```
This shouldn't be necessary given that you overrode the appropriate settings in the dynamically generated config file. Same goes for the `--catalog-url`.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+
+        self.validate_output()
+
+        # Drop the table afterwards so that the next test starts with a clean table.
```
This shouldn't be necessary since each test run calls `reset()` at the beginning which drops everything.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+            total_count = cursor.fetchone()[0]
+
+            expected_total_count = 6
+            self.assertEqual(total_count, expected_total_count)
```
I'd prefer this read in the expected data and then called `len()` on it to figure out the expected count. That way someone need only change the expected data file without having to also update this code if they add rows.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+                    for col_num in range(1, 6):
+                        found_datum = subject_listing[col_num]
+                        expected_datum = expected_row[col_num - 1]  # We subtract 1 since we removed the row number.
+                        if col_num == 2:  # Pandas doesn't parse the dates as datetime.date objects, so manually cast.
```
`read_csv()` takes in a `parse_dates` parameter which I believe can be used to solve this problem.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+            subjects = cursor.fetchall()
+
+            try:
+                for subject_listing in subjects:
```
I would prefer to modify the ingest mechanism such that we end up with two supposedly equal `DataFrame` objects that we can just assert are equal.

For example, you could explicitly list the columns read in the SQL statement so that you don't have to shift values over etc.

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+        working = True
+        self.assertGreater(data.shape[0], row_num)
+        row = data.iloc[row_num]
+        for expected_key in expected.keys():
```
Could we just do:

```python
self.assertEquals(dict(row), expected)
```

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+        data = self.run_task("{}")
+        self.assertEquals(data.shape[0], 0)
+
+    def test_course_no_sbujects(self):
```
nit: typo here -> s/sbujects/subjects/

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+    def test_course_no_sbujects(self):
+        """With a course with no subjects, we expect a row with NULLs."""
+
+        data = self.run_task("{\"items\":[{\"course_id\":\"foo\", \"subjects\":[{}]}]}")
```
I'd be OK with using `json.dumps()` and native python dict, list objects here to avoid messy backslash land.

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+        self.assertEquals(data.shape[0], 1)
+        # We expect nulls for the three subject data columns.
+        expected = {'course_id': 'foo',
+                    'subject_uri': '\N', 'subject_title': '\N', 'subject_language': '\N'}  # pylint: disable-msg=W1402
```
I typically just use `'\\N'` to avoid the pylint warning. If you decide not to do that, that's OK too, but we typically use the symbolic name when disabling pylint messages since it's a lot more clear to readers what exactly is being disabled, in this case it would be "anomalous-unicode-escape-in-string".

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+from mock import MagicMock
+
+
+class TestCourseSubjects(unittest.TestCase):
```
We should have a test for the corner case when the API returns a list instead of the expected dictionary.

Also for the case when the API returns invalid JSON, and/or unexpected keys / missing keys etc.

mulby commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+        self.temp_rootdir = tempfile.mkdtemp()
+        self.input_dir = os.path.join(self.temp_rootdir, "input")
+        os.mkdir(self.input_dir)
+        self.input_file = os.path.join(self.input_dir, "catalog", "20150625", "catalog.json")
```
This date should be 'dt=2015-06-25' for consistency with the rest of the pipeline and potential future Hive compatibility.

jab5569 replied:
url_path_join() instead?

jrowan replied:
I use os.path.join in these cases since the test data is stored locally, in the repo (e.g. in tests/acceptance/fixtures) or we're making temporary directories to store temporary files, and it's not worth putting them on s3.  I use url_path_join for all s3 or internet locations and os.path.join for all local locations, as in general os path separators and url path separators could be different.

mulby commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+"""
+Ensure we can write to MySQL data sources.
```
nit: s/MySQL/Vertica/

mulby replied:
(throughout this file)

jab5569 replied:
'Vertica' data source?

mulby commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+
+    def test_run(self):
+        self.create_task().run()
+        self.assertTrue(self.mock_vertica_connector.connect().cursor().execute.called)
```
I will often assign a core mock to a variable to make this type of thing slightly less verbose:

```python
mock_conn = self.mock_vertica_connector.connect()
self.assertFalse(mock_conn.rollback.called)
```

mulby commented:
finished my pass, getting closer!

jab5569 commented on edx/analytics/tasks/tests/acceptance/services/vertica.py:
```diff
@@ -0,0 +1,70 @@
+        """The credentials for connecting to the database, read from a URL."""
+        if not hasattr(self, '_credentials'):
+            target = get_target_from_url(self.vertica_creds_url)
+            print "TARGET BUILT: ", target
```
Does this need to be removed?

jab5569 commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,151 @@
+    Catalog with one course listed which has one subject
+    Catalog with one course listed which has multiple subjects
+    Catalog with multiple courses listed
+    Catalog that is an actual day's catalog, checking that the size of the produced .tsv is good.
```
Very nice

jab5569 commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+
+class VerticaCopyTaskTest(unittest.TestCase):
+    """
+    Ensure we can connect to and write data to MySQL data sources.
```
'Vertica'

jab5569 commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+
+    def create_task(self, credentials=None, source=None, overwrite=False, cls=CopyToVerticaDummyTable):
+        """
+         Emulate execution of a generic MysqlTask.
```
Vertica

jab5569 commented on edx/analytics/tasks/user_activity.py:
```diff
@@ -116,18 +129,15 @@ def reducer(self, key, values):
+        super(UserActivityTask, self).run()
 
 
-class UserActivityTableTask(UserActivityDownstreamMixin, HiveTableTask):
```
Why don't we need to instantiate UserActivityDownstreamMixin here?

jab5569 commented on edx/analytics/tasks/util/overwrite.py:
```diff
@@ -51,6 +51,6 @@ def remove_output_on_overwrite(self):
         if self.overwrite:
             self.attempted_removal = True
-            if self.output().exists():
+            if self.output().exists() and hasattr(self.output(), 'remove'):
```
What does this do?

jab5569 commented on edx/analytics/tasks/util/vertica_target.py:
```diff
@@ -0,0 +1,131 @@
+        """
+
+        # vertica-python 0.5.0 changes the code for connecting to databases to use kwargs instead of a dictionary.
+        # The 'database' parameter is included for DBAPI reasons and does not actually affect the session.
```
This differentiation is impressive

jab5569 commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+            self.assertEqual(total_count, expected_total_count)
+
+            # Verify that all the subject data loaded into Vertica is correct.
+            expected_output_csv = os.path.join(self.data_dir, 'output', 'expected_subjects_for_acceptance.csv')
```
url_path_join() instead?

jab5569 commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,89 @@
+
+    def upload_data(self):
+        """Puts the test course catalog where the processing task would look for it, bypassing calling the actual ALI"""
+        src = os.path.join(self.data_dir, 'input', self.INPUT_FILE)
```
url_path_join() instead?

jab5569 commented:
I'll defer to @mulby  for any showstoppers, what I found were mostly nits and questions.

Overall this is great work James; very impressive. 

:+1:

jrowan commented:
@mulby @jab5569 I've gone through and believe I have addressed your concerns.  I've added the additional unit tests Gabe suggested, the streamlined unit and acceptance tests are passing, and I'm getting a mostly* clean bill of code cleanliness.  Sorry for the massive quantity of commits that arose when I was trying to debug an acceptance test; I will of course rebase before merging.

\* nitty NB: the creation of mock course catalog json dictionaries in test_course_subjects.py leads to situations that, at least according to PyCharm, are "pep8 paradoxes."  There are some lines marked as "continuation line is over-indented for visual indent" that, when I go through space by space trying to fix the indentation level, go immediately from "under-indented" to "over-indented," without every passing through a pep8 acceptable state.

mulby commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,69 @@
+                                                                    'subject_title', 'subject_language'])
+
+            # The equality check gives a pandas dataframe object by default, so we condense it to a single boolean.
+            self.assertTrue(all(subjects == expected))
```
Nit: instead of checking the count and then row equality... you should just be able to use:

```python
from pandas.util.testing import assert_frame_equal
assert_frame_equal(data_frame_a, data_frame_b)
```

mulby commented on edx/analytics/tasks/tests/acceptance/__init__.py:
```diff
@@ -94,6 +98,13 @@ def setUp(self):
+                'schema': schema
+            },
+            'course-catalog': {
+                'catalog_path': self.catalog_path
```
Nit: note that by default the acceptance tests use the `config/test.cfg` file, so you shouldn't have to put in this override if you define this static value in that file (which you do above). These dynamic overrides are only here since there are many values that are not static and have to be overwritten at runtime.

mulby commented:
Great work @jrowan - let's get this on master!

:+1:

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,178 @@
+    """
+    A task that reads a local file generated from a daily catalog pull, and writes the course id and subject to a tsv.
+
+
```
nit: two blank lines seems excessive....

brianhw commented on edx/analytics/tasks/course_catalog.py:
```diff
@@ -0,0 +1,178 @@
+                for course in catalog:
+                    # The course catalog occasionally is buggy and has malformed courses, so just skip those.
+                    if not type(course) == dict:
+                        continue
```
Should we at least output a warning, so we have a way of determining how often this occurs over time?

brianhw commented on edx/analytics/tasks/tests/acceptance/test_course_catalog.py:
```diff
@@ -0,0 +1,71 @@
+
+            # The equality check gives a pandas dataframe object by default, so we condense it to a single boolean.
+            assert_frame_equal(subjects, expected)
+            # self.assertTrue(all(subjects == expected))
```
this can be removed now...

brianhw commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,239 @@
+        """
+        Checks if the entries in a row of the data are what we expected, and returns whether this is true.
+
+
```
nit: extra blank line here and below.

brianhw commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,239 @@
+
+        Args:
+            data is a pandas data frame representing the output of a run of the task.
+            row_num is the row number to check.
```
nit: note if it's zero or one-based...

brianhw commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,239 @@
+        # We expect an entry in the list of courses, since there is a course in the catalog.
+        self.assertEquals(data.shape[0], 1)
+        # We expect nulls for the three subject data columns.
+        expected = {'course_id': 'foo', 'date': '2015-06-25',
```
nit:  'date' should be on a separate line, at least, and actually, so should 'course_id'.

brianhw commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,239 @@
+        data = self.run_task(json.dumps(input_data))
+        # We expect to see this course with the mock_subject information.
+        self.assertEquals(data.shape[0], 1)
+        expected = {'course_id': 'foo', 'subject_uri': 'testing/uri', 'subject_title': 'Testing', 'date': '2015-06-25',
```
nit: the date here should match the date used in the setup.  Pull it out into a constant to make that clear.

brianhw commented on edx/analytics/tasks/tests/test_course_subjects.py:
```diff
@@ -0,0 +1,239 @@
+                            "subjects":
+                                [{"uri": "testing/uri",
+                                  "title": "Testing",
+                                  "language": "py"}]}
```
nit: if it's a different course, it should presumably have different title and uri....

brianhw commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+    def test_create_table_without_column_definition(self):
+        connection = MagicMock()
+        task = self.create_task(cls=CopyToPredefinedVerticaDummyTable)
+        with self.assertRaises(NotImplementedError):
```
nit: should this also check the error message to clarify the right error is occurring?  (e.g table vs column).  There's a regexp version of assertRaises that does this...

brianhw commented on edx/analytics/tasks/tests/test_vertica_load.py:
```diff
@@ -0,0 +1,226 @@
+        cursor = MagicMock()
+        task.copy_data_table_from_target(cursor)
+        query = cursor.copy_file.call_args[0][0]
+        self.assertEquals(query, self._get_expected_query())
```
get_expected_query is called here with only one row, but _get_source_string was called with 4.  What happened to the other three?

jrowan replied:
Since we're using Vertica copies, we don't need to make 4 queries.  The set of all 4 rows is used when we compare the readout of the file object marked as the insert source to the readout of the file object output by the copy command; all 4 rows get returned by call_args[0][1].

brianhw commented on edx/analytics/tasks/vertica_load.py:
```diff
@@ -0,0 +1,285 @@
+
+        credentials: Path to the external access credentials file.
+        schema:  The schema to which to write.
+        insert_chunk_size:  The number of rows to insert at a time.
```
where is this used?  cut-and-paste error?

brianhw commented:
LGTM. :+1:
